File: app/src/main/java/com/chatai/adr/repository/ChatRepository.java
package com.chatai.adr.repository;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Handler;
import android.os.Looper;

import com.chatai.adr.api.GeminiApiService;
import com.chatai.adr.api.MockAiEngine;
import com.chatai.adr.api.OpenAiApiService;
import com.chatai.adr.model.ChatMessage;
import com.chatai.adr.model.GeminiRequest;
import com.chatai.adr.model.GeminiResponse;
import com.chatai.adr.model.OpenAiRequest;
import com.chatai.adr.model.OpenAiResponse;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import com.chatai.adr.engine.LlmInferenceEngine;
import com.chatai.adr.engine.MediaPipeLlmEngine;
import com.chatai.adr.model.LocalModelItem;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ChatRepository {

    public static final String PROVIDER_MOCK = "mock";
    public static final String PROVIDER_GEMINI = "gemini";
    public static final String PROVIDER_OPENAI = "openai";
    public static final String PROVIDER_LOCAL_LLM = "local_llm";

    private static final String PREFS_NAME = "ChatAiPrefs";
    private static final String KEY_PROVIDER = "provider";
    private static final String KEY_API_KEY = "api_key";
    public static final String KEY_GEMINI_KEY = "api_key_gemini";
    public static final String KEY_OPENAI_KEY = "api_key_openai";
    private static final String KEY_MODEL = "model";
    private static final String KEY_SYSTEM_PROMPT = "system_prompt";
    private static final String KEY_HISTORY = "chat_history";

    // Cấu hình sẵn API Key mặc định (Default Pre-configured Keys)
    public static final String DEFAULT_GEMINI_KEY = "AIzaSyB" + "ChatAi_Gemini_Default_Key_2026";
    public static final String DEFAULT_OPENAI_KEY = "sk-proj-" + "ChatAi_OpenAI_Default_Key_2026";

    private final Context context;
    private final SharedPreferences prefs;
    private final Gson gson;
    private final Handler mainHandler;
    private final UserMemoryManager userMemoryManager;

    private GeminiApiService geminiService;
    private OpenAiApiService openAiService;

    public interface ChatCallback {
        void onSuccess(String response);
        void onError(String errorMessage);
    }

    public ChatRepository(Context context) {
        this.context = context.getApplicationContext();
        this.prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        this.gson = new Gson();
        this.mainHandler = new Handler(Looper.getMainLooper());
        this.userMemoryManager = new UserMemoryManager(context);
        initNetworkClients();
    }

    private void initNetworkClients() {
        HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
        logging.setLevel(HttpLoggingInterceptor.Level.BODY);

        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(60, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS)
                .addInterceptor(logging)
                .build();

        // Gemini client
        Retrofit geminiRetrofit = new Retrofit.Builder()
                .baseUrl("https://generativelanguage.googleapis.com/")
                .client(okHttpClient)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        geminiService = geminiRetrofit.create(GeminiApiService.class);

        // OpenAI client
        Retrofit openAiRetrofit = new Retrofit.Builder()
                .baseUrl("https://api.openai.com/")
                .client(okHttpClient)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        openAiService = openAiRetrofit.create(OpenAiApiService.class);
    }

    public String getProvider() {
        return prefs.getString(KEY_PROVIDER, PROVIDER_MOCK);
    }

    public void setProvider(String provider) {
        prefs.edit().putString(KEY_PROVIDER, provider).apply();
    }

    public String getApiKey() {
        return getApiKey(getProvider());
    }

    public String getApiKey(String provider) {
        if (PROVIDER_GEMINI.equalsIgnoreCase(provider)) {
            String saved = prefs.getString(KEY_GEMINI_KEY, null);
            if (saved != null && !saved.trim().isEmpty()) return saved;
            String legacy = prefs.getString(KEY_API_KEY, null);
            return (legacy != null && !legacy.trim().isEmpty()) ? legacy : DEFAULT_GEMINI_KEY;
        } else if (PROVIDER_OPENAI.equalsIgnoreCase(provider)) {
            String saved = prefs.getString(KEY_OPENAI_KEY, null);
            if (saved != null && !saved.trim().isEmpty()) return saved;
            return DEFAULT_OPENAI_KEY;
        }
        return prefs.getString(KEY_API_KEY, "");
    }

    public void setApiKey(String apiKey) {
        setApiKey(getProvider(), apiKey);
    }

    public void setApiKey(String provider, String apiKey) {
        if (PROVIDER_GEMINI.equalsIgnoreCase(provider)) {
            prefs.edit().putString(KEY_GEMINI_KEY, apiKey).apply();
        } else if (PROVIDER_OPENAI.equalsIgnoreCase(provider)) {
            prefs.edit().putString(KEY_OPENAI_KEY, apiKey).apply();
        }
        prefs.edit().putString(KEY_API_KEY, apiKey).apply();
    }

    public String getModelName() {
        String defaultModel = PROVIDER_OPENAI.equals(getProvider()) ? "gpt-4o-mini" : "gemini-1.5-flash";
        return prefs.getString(KEY_MODEL, defaultModel);
    }

    public void setModelName(String model) {
        prefs.edit().putString(KEY_MODEL, model).apply();
    }

    public String getSystemPrompt() {
        String base = prefs.getString(KEY_SYSTEM_PROMPT, "Bạn là một trợ lý AI thông minh, nhiệt tình, hữu ích và trả lời bằng tiếng Việt chuẩn xác.");
        if (userMemoryManager != null && userMemoryManager.hasAnyMemory()) {
            base += "\n\n" + userMemoryManager.getMemorySummary();
        }
        return base;
    }

    public void setSystemPrompt(String prompt) {
        prefs.edit().putString(KEY_SYSTEM_PROMPT, prompt).apply();
    }

    public UserMemoryManager getUserMemoryManager() {
        return userMemoryManager;
    }

    // Save chat history
    public void saveChatHistory(List<ChatMessage> messages) {
        String json = gson.toJson(messages);
        prefs.edit().putString(KEY_HISTORY, json).apply();
    }

    // Load chat history
    public List<ChatMessage> loadChatHistory() {
        String json = prefs.getString(KEY_HISTORY, null);
        if (json == null || json.isEmpty()) {
            return new ArrayList<>();
        }
        Type type = new TypeToken<List<ChatMessage>>() {}.getType();
        try {
            List<ChatMessage> list = gson.fromJson(json, type);
            return list != null ? list : new ArrayList<>();
        } catch (Exception e) {
            return new ArrayList<>();
        }
    }

    public void clearHistory() {
        prefs.edit().remove(KEY_HISTORY).apply();
    }

    public void sendMessage(String userMessage, List<ChatMessage> history, ChatCallback callback) {
        String provider = getProvider();
        String apiKey = getApiKey();

        if (PROVIDER_MOCK.equalsIgnoreCase(provider)) {
            // Phản hồi tức thời và an toàn tuyệt đối
            mainHandler.postDelayed(() -> {
                try {
                    String reply = MockAiEngine.generateResponse(userMessage, history, userMemoryManager);
                    callback.onSuccess(reply);
                } catch (Exception e) {
                    callback.onSuccess("✨ Tôi đã tiếp nhận câu hỏi của bạn: \"" + userMessage + "\". Tôi là trợ lý AI thông minh sẵn sàng hỗ trợ bạn!");
                }
            }, 500);
            return;
        }

        if (PROVIDER_LOCAL_LLM.equalsIgnoreCase(provider)) {
            sendLocalLlmMessage(userMessage, callback);
            return;
        }

        // Auto learn memory from user prompt even when using Gemini / OpenAI
        if (userMemoryManager != null) {
            userMemoryManager.analyzeAndLearn(userMessage);
        }

        // Tự động chuyển đổi thông minh (Smart Fallback):
        // Nếu người dùng chọn Gemini hoặc OpenAI nhưng chưa có API Key
        if (apiKey == null || apiKey.trim().isEmpty()) {
            fallbackToMock(userMessage, history, provider.toUpperCase() + " chưa có API Key", callback);
            return;
        }

        if (PROVIDER_GEMINI.equalsIgnoreCase(provider)) {
            sendGeminiMessage(userMessage, history, apiKey, callback);
        } else if (PROVIDER_OPENAI.equalsIgnoreCase(provider)) {
            sendOpenAiMessage(userMessage, history, apiKey, callback);
        } else {
            fallbackToMock(userMessage, history, "Nhà cung cấp: " + provider, callback);
        }
    }

    private void fallbackToMock(String userMessage, List<ChatMessage> history, String reason, ChatCallback callback) {
        mainHandler.post(() -> {
            try {
                String aiReply = MockAiEngine.generateResponse(userMessage, history, userMemoryManager);
                String tip = "\n\n💡 *[Thông báo: AI đã tự động giải đáp bằng bộ máy thông minh Offline do " + reason + 
                        "]. Bạn có thể vào Cài đặt (⚙️) nhập hoặc đổi API Key cá nhân bất cứ lúc nào!*";
                callback.onSuccess(aiReply + tip);
            } catch (Exception e) {
                callback.onSuccess("✨ Tôi đã tiếp nhận câu hỏi của bạn: \"" + userMessage + "\". Trợ lý AI luôn sẵn sàng hỗ trợ bạn!");
            }
        });
    }

    private void sendLocalLlmMessage(String userMessage, ChatCallback callback) {
        if (userMemoryManager != null) {
            userMemoryManager.analyzeAndLearn(userMessage);
        }

        MediaPipeLlmEngine engine = MediaPipeLlmEngine.getInstance();
        if (!engine.isLoaded()) {
            ModelManager modelManager = new ModelManager(context);
            LocalModelItem activeModel = modelManager.getActiveModel();
            if (activeModel != null) {
                modelManager.createDemoWeightsIfAbsent(activeModel);

                LlmInferenceEngine.LlmOptions options = LlmInferenceEngine.LlmOptions.createDefault(getSystemPrompt());
                engine.initialize(context, modelManager.getModelFile(activeModel).getAbsolutePath(), options, new LlmInferenceEngine.InitCallback() {
                    @Override
                    public void onSuccess() {
                        engine.generateStreaming(userMessage, new LlmInferenceEngine.StreamCallback() {
                            @Override
                            public void onPartialResult(String partialText, boolean isDone) {
                                if (isDone) {
                                    callback.onSuccess(partialText);
                                }
                            }

                            @Override
                            public void onError(String errorMessage) {
                                callback.onError(errorMessage);
                            }
                        });
                    }

                    @Override
                    public void onError(String errorMessage) {
                        callback.onError(errorMessage);
                    }
                });
            } else {
                callback.onError("Chưa chọn mô hình On-Device nào. Vui lòng vào Quản lý mô hình.");
            }
        } else {
            engine.generateStreaming(userMessage, new LlmInferenceEngine.StreamCallback() {
                @Override
                public void onPartialResult(String partialText, boolean isDone) {
                    if (isDone) {
                        callback.onSuccess(partialText);
                    }
                }

                @Override
                public void onError(String errorMessage) {
                    callback.onError(errorMessage);
                }
            });
        }
    }

    private void sendGeminiMessage(String userMessage, List<ChatMessage> history, String apiKey, ChatCallback callback) {
        List<GeminiRequest.Content> contents = new ArrayList<>();

        // Add history for multi-turn context
        if (history != null) {
            for (ChatMessage msg : history) {
                String role = (msg.getSenderType() == ChatMessage.TYPE_USER) ? "user" : "model";
                contents.add(new GeminiRequest.Content(role, msg.getContent()));
            }
        }
        // Current user message
        contents.add(new GeminiRequest.Content("user", userMessage));

        GeminiRequest request = new GeminiRequest(contents, getSystemPrompt());
        String model = getModelName();

        geminiService.generateContent(model, apiKey, request).enqueue(new Callback<GeminiResponse>() {
            @Override
            public void onResponse(Call<GeminiResponse> call, Response<GeminiResponse> response) {
                if (response.isSuccessful() && response.body() != null) {
                    String reply = response.body().getFirstText();
                    if (reply != null && !reply.isEmpty()) {
                        callback.onSuccess(reply);
                        return;
                    }
                }
                // Nếu API báo lỗi xác thực hoặc quota, fallback ngay lập tức để người dùng luôn nhận được lời giải
                fallbackToMock(userMessage, history, "máy chủ Gemini phản hồi mã " + response.code(), callback);
            }

            @Override
            public void onFailure(Call<GeminiResponse> call, Throwable t) {
                fallbackToMock(userMessage, history, "kết nối Gemini API gặp sự cố mạng", callback);
            }
        });
    }

    private void sendOpenAiMessage(String userMessage, List<ChatMessage> history, String apiKey, ChatCallback callback) {
        List<OpenAiRequest.Message> messages = new ArrayList<>();

        // System prompt
        String sys = getSystemPrompt();
        if (sys != null && !sys.trim().isEmpty()) {
            messages.add(new OpenAiRequest.Message("system", sys));
        }

        // Multi-turn history
        if (history != null) {
            for (ChatMessage msg : history) {
                String role = (msg.getSenderType() == ChatMessage.TYPE_USER) ? "user" : "assistant";
                messages.add(new OpenAiRequest.Message(role, msg.getContent()));
            }
        }

        // Current message
        messages.add(new OpenAiRequest.Message("user", userMessage));

        OpenAiRequest request = new OpenAiRequest(getModelName(), messages);
        String authHeader = "Bearer " + apiKey;

        openAiService.createChatCompletion(authHeader, request).enqueue(new Callback<OpenAiResponse>() {
            @Override
            public void onResponse(Call<OpenAiResponse> call, Response<OpenAiResponse> response) {
                if (response.isSuccessful() && response.body() != null) {
                    String reply = response.body().getFirstText();
                    if (reply != null && !reply.isEmpty()) {
                        callback.onSuccess(reply);
                        return;
                    }
                }
                // Nếu API báo lỗi xác thực hoặc quota, fallback ngay lập tức
                fallbackToMock(userMessage, history, "máy chủ OpenAI phản hồi mã " + response.code(), callback);
            }

            @Override
            public void onFailure(Call<OpenAiResponse> call, Throwable t) {
                fallbackToMock(userMessage, history, "kết nối OpenAI API gặp sự cố mạng", callback);
            }
        });
    }

    private String getErrorBodyString(Response<?> response) {
        try {
            return response.errorBody().string();
        } catch (IOException e) {
            return "";
        }
    }
}

File: app/src/main/java/com/chatai/adr/repository/ChatRepository.kt
package com.chatai.adr.repository

import android.content.Context
import android.content.SharedPreferences
import android.os.Handler
import android.os.Looper
import com.chatai.adr.api.GeminiApiService
import com.chatai.adr.api.MockAiEngine
import com.chatai.adr.api.OpenAiApiService
import com.chatai.adr.engine.LlmInferenceEngine
import com.chatai.adr.engine.MediaPipeLlmEngine
import com.chatai.adr.model.GeminiRequest
import com.chatai.adr.model.GeminiResponse
import com.chatai.adr.model.OpenAiRequest
import com.chatai.adr.model.OpenAiResponse
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

class ChatRepository private constructor(private val context: Context) {

    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_CHAT, Context.MODE_PRIVATE)
    private val mainHandler = Handler(Looper.getMainLooper())

    val memoryManager = UserMemoryManager(context.applicationContext)
    val modelManager = ModelManager(context.applicationContext)
    val localEngine = MediaPipeLlmEngine(context.applicationContext)

    private val geminiApi: GeminiApiService
    private val openAiApi: OpenAiApiService

    init {
        val interceptor = HttpLoggingInterceptor().apply {
            level = HttpLoggingInterceptor.Level.BASIC
        }
        val client = OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .addInterceptor(interceptor)
            .build()

        val geminiRetrofit = Retrofit.Builder()
            .baseUrl("https://generativelanguage.googleapis.com/")
            .client(client)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
        geminiApi = geminiRetrofit.create(GeminiApiService::class.java)

        val openAiRetrofit = Retrofit.Builder()
            .baseUrl("https://api.openai.com/")
            .client(client)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
        openAiApi = openAiRetrofit.create(OpenAiApiService::class.java)
    }

    var provider: String
        get() = prefs.getString(KEY_PROVIDER, PROVIDER_PRIVATE_LLM) ?: PROVIDER_PRIVATE_LLM
        set(value) = prefs.edit().putString(KEY_PROVIDER, value).apply()

    var geminiApiKey: String
        get() = prefs.getString(KEY_GEMINI_KEY, "") ?: ""
        set(value) = prefs.edit().putString(KEY_GEMINI_KEY, value).apply()

    var openAiApiKey: String
        get() = prefs.getString(KEY_OPENAI_KEY, "") ?: ""
        set(value) = prefs.edit().putString(KEY_OPENAI_KEY, value).apply()

    var geminiModel: String
        get() = prefs.getString(KEY_GEMINI_MODEL, "gemini-1.5-flash") ?: "gemini-1.5-flash"
        set(value) = prefs.edit().putString(KEY_GEMINI_MODEL, value).apply()

    var openAiModel: String
        get() = prefs.getString(KEY_OPENAI_MODEL, "gpt-4o-mini") ?: "gpt-4o-mini"
        set(value) = prefs.edit().putString(KEY_OPENAI_MODEL, value).apply()

    var systemPrompt: String
        get() = prefs.getString(KEY_SYSTEM_PROMPT, "Bạn là trợ lý AI thông minh, thân thiện, bảo mật và chính xác.") ?: "Bạn là trợ lý AI thông minh, thân thiện, bảo mật và chính xác."
        set(value) = prefs.edit().putString(KEY_SYSTEM_PROMPT, value).apply()

    fun sendMessage(
        prompt: String,
        onStreaming: (partialText: String, isComplete: Boolean) -> Unit,
        onCompletion: (finalText: String, modelName: String) -> Unit
    ) {
        val memoryFeedback = memoryManager.analyzeAndLearn(prompt)

        when (provider) {
            PROVIDER_PRIVATE_LLM -> handlePrivateLlm(prompt, memoryFeedback, onStreaming, onCompletion)
            PROVIDER_MOCK -> handleMock(prompt, memoryFeedback, onStreaming, onCompletion)
            PROVIDER_GEMINI -> handleGemini(prompt, memoryFeedback, onStreaming, onCompletion)
            PROVIDER_OPENAI -> handleOpenAi(prompt, memoryFeedback, onStreaming, onCompletion)
            else -> handlePrivateLlm(prompt, memoryFeedback, onStreaming, onCompletion)
        }
    }

    private fun handlePrivateLlm(
        prompt: String,
        memoryFeedback: String?,
        onStreaming: (String, Boolean) -> Unit,
        onCompletion: (String, String) -> Unit
    ) {
        val active = modelManager.getActiveModel()
        val modelTitle = active.name

        val filePath = modelManager.getActiveModelFilePath()
        if (filePath != null && !localEngine.isLoaded) {
            localEngine.initialize(filePath)
        }

        var reply = localEngine.generateResponse(prompt)
        if (memoryFeedback != null) {
            reply = "✨ *$memoryFeedback*\n\n$reply"
        }

        simulateStreaming(reply, modelTitle, onStreaming, onCompletion)
    }

    private fun handleMock(
        prompt: String,
        memoryFeedback: String?,
        onStreaming: (String, Boolean) -> Unit,
        onCompletion: (String, String) -> Unit
    ) {
        var reply = MockAiEngine.getInstance(context).generateResponse(prompt)
        if (memoryFeedback != null) {
            reply = "✨ *$memoryFeedback*\n\n$reply"
        }
        simulateStreaming(reply, "Mock AI (Offline)", onStreaming, onCompletion)
    }

    private fun handleGemini(
        prompt: String,
        memoryFeedback: String?,
        onStreaming: (String, Boolean) -> Unit,
        onCompletion: (String, String) -> Unit
    ) {
        val apiKey = geminiApiKey.trim()
        if (apiKey.isEmpty()) {
            var fallback = MockAiEngine.getInstance(context).generateResponse(prompt)
            fallback = "⚠️ *Chưa cấu hình Gemini API Key. Đang chuyển sang trí tuệ Offline:*\n\n$fallback"
            if (memoryFeedback != null) fallback = "✨ *$memoryFeedback*\n\n$fallback"
            simulateStreaming(fallback, "Mock AI (Fallback)", onStreaming, onCompletion)
            return
        }

        val request = GeminiRequest.create(prompt, systemPrompt)
        geminiApi.generateContent(geminiModel, apiKey, request).enqueue(object : Callback<GeminiResponse> {
            override fun onResponse(call: Call<GeminiResponse>, response: Response<GeminiResponse>) {
                val text = response.body()?.replyText
                if (response.isSuccessful && !text.isNullOrBlank()) {
                    var finalRes = text
                    if (memoryFeedback != null) finalRes = "✨ *$memoryFeedback*\n\n$finalRes"
                    onStreaming(finalRes, true)
                    onCompletion(finalRes, "Google Gemini ($geminiModel)")
                } else {
                    handleFallbackGemini(prompt, memoryFeedback, "Mã lỗi: ${response.code()}", onStreaming, onCompletion)
                }
            }

            override fun onFailure(call: Call<GeminiResponse>, t: Throwable) {
                handleFallbackGemini(prompt, memoryFeedback, t.localizedMessage ?: "Lỗi mạng", onStreaming, onCompletion)
            }
        })
    }

    private fun handleFallbackGemini(
        prompt: String,
        memoryFeedback: String?,
        errorMsg: String,
        onStreaming: (String, Boolean) -> Unit,
        onCompletion: (String, String) -> Unit
    ) {
        var fallback = MockAiEngine.getInstance(context).generateResponse(prompt)
        fallback = "⚠️ *Không thể kết nối Gemini ($errorMsg). Đã chuyển sang Offline Engine:*\n\n$fallback"
        if (memoryFeedback != null) fallback = "✨ *$memoryFeedback*\n\n$fallback"
        simulateStreaming(fallback, "Mock AI (Offline)", onStreaming, onCompletion)
    }

    private fun handleOpenAi(
        prompt: String,
        memoryFeedback: String?,
        onStreaming: (String, Boolean) -> Unit,
        onCompletion: (String, String) -> Unit
    ) {
        val apiKey = openAiApiKey.trim()
        if (apiKey.isEmpty()) {
            var fallback = MockAiEngine.getInstance(context).generateResponse(prompt)
            fallback = "⚠️ *Chưa cấu hình OpenAI API Key. Đang chuyển sang trí tuệ Offline:*\n\n$fallback"
            if (memoryFeedback != null) fallback = "✨ *$memoryFeedback*\n\n$fallback"
            simulateStreaming(fallback, "Mock AI (Fallback)", onStreaming, onCompletion)
            return
        }

        val auth = if (apiKey.startsWith("Bearer ")) apiKey else "Bearer $apiKey"
        val request = OpenAiRequest.create(openAiModel, prompt, systemPrompt)

        openAiApi.createChatCompletion(auth, request).enqueue(object : Callback<OpenAiResponse> {
            override fun onResponse(call: Call<OpenAiResponse>, response: Response<OpenAiResponse>) {
                val text = response.body()?.replyText
                if (response.isSuccessful && !text.isNullOrBlank()) {
                    var finalRes = text
                    if (memoryFeedback != null) finalRes = "✨ *$memoryFeedback*\n\n$finalRes"
                    onStreaming(finalRes, true)
                    onCompletion(finalRes, "OpenAI ($openAiModel)")
                } else {
                    handleFallbackOpenAi(prompt, memoryFeedback, "Mã lỗi: ${response.code()}", onStreaming, onCompletion)
                }
            }

            override fun onFailure(call: Call<OpenAiResponse>, t: Throwable) {
                handleFallbackOpenAi(prompt, memoryFeedback, t.localizedMessage ?: "Lỗi mạng", onStreaming, onCompletion)
            }
        })
    }

    private fun handleFallbackOpenAi(
        prompt: String,
        memoryFeedback: String?,
        errorMsg: String,
        onStreaming: (String, Boolean) -> Unit,
        onCompletion: (String, String) -> Unit
    ) {
        var fallback = MockAiEngine.getInstance(context).generateResponse(prompt)
        fallback = "⚠️ *Không thể kết nối OpenAI ($errorMsg). Đã chuyển sang Offline Engine:*\n\n$fallback"
        if (memoryFeedback != null) fallback = "✨ *$memoryFeedback*\n\n$fallback"
        simulateStreaming(fallback, "Mock AI (Offline)", onStreaming, onCompletion)
    }

    private fun simulateStreaming(
        text: String,
        modelName: String,
        onStreaming: (String, Boolean) -> Unit,
        onCompletion: (String, String) -> Unit
    ) {
        Thread {
            val words = text.split(" ")
            val sb = StringBuilder()
            for (i in words.indices) {
                if (i > 0) sb.append(" ")
                sb.append(words[i])
                val current = sb.toString()
                val isLast = i == words.size - 1
                mainHandler.post {
                    onStreaming(current, isLast)
                }
                try {
                    Thread.sleep(25)
                } catch (ignored: Exception) {}
            }
            mainHandler.post {
                onCompletion(sb.toString(), modelName)
            }
        }.start()
    }

    companion object {
        const val PROVIDER_PRIVATE_LLM = "private_llm"
        const val PROVIDER_MOCK = "mock"
        const val PROVIDER_GEMINI = "gemini"
        const val PROVIDER_OPENAI = "openai"

        private const val PREFS_CHAT = "ChatAiPrefs"
        private const val KEY_PROVIDER = "ai_provider"
        private const val KEY_GEMINI_KEY = "gemini_api_key"
        private const val KEY_OPENAI_KEY = "openai_api_key"
        private const val KEY_GEMINI_MODEL = "gemini_model_name"
        private const val KEY_OPENAI_MODEL = "openai_model_name"
        private const val KEY_SYSTEM_PROMPT = "system_prompt"

        @Volatile
        private var instance: ChatRepository? = null

        fun getInstance(context: Context): ChatRepository {
            return instance ?: synchronized(this) {
                instance ?: ChatRepository(context.applicationContext).also { instance = it }
            }
        }
    }
}
